File: yabte/backtest/book.py
import logging
from collections import defaultdict
from dataclasses import dataclass, field
from decimal import Decimal
from itertools import groupby
from typing import Any, Dict, List, Sequence, TypeAlias

import pandas as pd

from ._helpers import ensure_decimal
from .asset import Asset, AssetName
from .transaction import CashTransaction, Trade, Transaction

logger = logging.getLogger(__name__)


__all__ = ["Book"]


@dataclass(kw_only=True)
class BookMandate:
    def check(self, current_pos, quantity) -> bool:
        raise NotImplementedError()


BookName: TypeAlias = str
"""Book name string."""


@dataclass(kw_only=True)
class Book:
    """Record of asset trades and positions including cash.

    A `name` can be provided and the default `cash` value can be changed
    to a non-zero amount.
    """

    name: BookName
    """Name of book."""

    mandates: Dict[AssetName, BookMandate] = field(default_factory=dict)
    """Dictionary mapping assets to mandates (experimental)."""

    positions: Dict[AssetName, Decimal] = field(
        default_factory=lambda: defaultdict(Decimal)
    )
    """Dictionary tracking positions of each asset."""

    transactions: List[Transaction] = field(default_factory=list)
    """List of executed trades."""

    denom: str = "USD"
    """Book currency."""

    cash: Decimal = Decimal(0)
    """Cash value of book."""

    rate: Decimal = Decimal(0)
    """Daily interest rate applied to cash at end of day."""

    interest_round_dp: int = 3
    """Number of decimal places to round interest."""

    _history: List[List[Any]] = field(default_factory=list)

    @property
    def history(self) -> pd.DataFrame:
        """Dataframe with book cash, mtm and total value history."""
        return pd.DataFrame(
            self._history, columns=["ts", "cash", "mtm", "total"]
        ).set_index("ts")

    def __post_init__(self):
        self.cash = ensure_decimal(self.cash)
        self.rate = ensure_decimal(self.rate)

    def test_trades(self, trades: Sequence[Trade]) -> bool:
        """Checks whether list of trades will be successful by not failing any
        mandates."""
        for asset_name, asset_trades in groupby(trades, lambda t: t.asset_name):
            if asset_name in self.mandates:
                total_quantity = sum(t.quantity for t in asset_trades)
                if not self.mandates[asset_name].check(
                    self.positions[asset_name], total_quantity
                ):
                    return False
        return True

    def add_transactions(self, transactions: Sequence[Transaction]):
        """Records the `transactions` and adjusts internal dictionary of
        positions and value of cash accordingly."""
        for tran in transactions:
            if isinstance(tran, Trade):
                self.positions[tran.asset_name] += tran.quantity
                self.cash += tran.total
            elif isinstance(tran, CashTransaction):
                self.cash += tran.total
            else:
                raise ValueError(f"Unsupport transaction class: {type(tran)}")

            self.transactions.append(tran)

    def eod_tasks(
        self, ts: pd.Timestamp, day_data: pd.DataFrame, asset_map: Dict[str, Asset]
    ):
        """Run end of day tasks such as book keeping."""
        # accumulate continously compounded interest
        interest = round(self.cash * (self.rate.exp() - 1), self.interest_round_dp)
        if self.rate != 0 and interest != 0:
            self.add_transactions(
                [
                    CashTransaction(
                        ts=ts,
                        total=interest,
                        desc=f"interest payment on cash {self.cash:.2f}",
                    )
                ]
            )
        cash = float(self.cash)
        mtm = sum(
            day_data[asset_map[an].data_label].Close * float(q)
            for an, q in self.positions.items()
        )
        self._history.append([ts, cash, mtm, cash + mtm])

File: cpp/src/YABTE/BackTest/Book.cpp
#include "YABTE/BackTest/Book.hpp"

#include <arrow/stl.h>

#include <cmath>

#include "YABTE/Utilities/Arrow/TableHelpers.hpp"

using std::make_shared;

namespace YABTE::BackTest {

Book::Book(const string& name, const string& denom, double cash, double rate,
           int interest_round_dp)
    : name_(name),
      denom_(denom),
      cash_(cash),
      rate_(rate),
      interest_round_dp_(interest_round_dp) {}

shared_ptr<Book> Book::clone() const { return make_shared<Book>(*this); }

bool Book::test_trades(const vector<shared_ptr<Trade>>& trades) const {
    // TODO: Implement this function
    // for asset_name, asset_trades in groupby(trades, lambda t: t.asset_name):
    //     if asset_name in self.mandates:
    //         total_quantity = sum(t.quantity for t in asset_trades)
    //         if not self.mandates[asset_name].check(
    //             self.positions[asset_name], total_quantity
    //         ):
    //             return False

    return true;
}

void Book::add_transactions(const TransactionVector& transactions) {
    for (auto& tran : transactions) {
        if (auto trade = dynamic_cast<const Trade*>(tran.get());
            trade != nullptr) {
            this->positions_[trade->asset_name_] += trade->quantity_;
            this->cash_ += trade->total_;
        } else if (auto ctran =
                       dynamic_cast<const CashTransaction*>(tran.get());
                   ctran != nullptr) {
            this->cash_ += ctran->total_;
        } else {
            throw std::runtime_error("Unsupport transaction class");
        }

        this->transactions_.push_back(tran);
    }
}

void Book::eod_tasks(const Timestamp& ts, const DayData& day_data,
                     const AssetMap& asset_map) {
    // Run end of day tasks such as book keeping."""
    // accumulate continously compounded interest

    auto interest = round_n_digits(this->cash_ * (std::exp(this->rate_) - 1),
                                   this->interest_round_dp_);
    if (this->rate_ != 0 && interest != 0) {
        auto cash_trans = CashTransaction(
            ts, interest, "interest payment on cash {self.cash:.2f}"s);
        const TransactionVector trans = {make_shared<Transaction>(cash_trans)};
        this->add_transactions(trans);
    }

    auto mtm = 0.0;
    for (const auto& [an, q] : this->positions_) {
        if (auto asset = asset_map.at(an); asset != nullptr) {
            mtm += asset->end_of_day_price(*asset->_filter_data(day_data)) * q;
        }
    }

    this->_history_.push_back({ts, this->cash_, mtm, this->cash_ + mtm});
    // cash = float(self.cash)
    // mtm = float(
    //     sum(
    //         asset.end_of_day_price(asset._filter_data(day_data)) * q
    //         for an, q in self.positions.items()
    //         if (asset := asset_map.get(an))
    //     )
    // )
    // self._history.append([ts, cash, mtm, cash + mtm])
}

shared_ptr<Table> Book::history() const {
    shared_ptr<Table> table;
    vector<std::string> names = {"ts", "cash", "mtm", "total"};
    if (!arrow::stl::TableFromTupleRange(arrow::default_memory_pool(),
                                         this->_history_, names, &table)
             .ok()) {
        throw std::runtime_error("Error creating table");
    }
    return table;
}

}  // namespace YABTE::BackTest